File: leetcode/Minimum_Path_Sum.java
import java.util.*;

public class Solution {
    public int minPathSum(int[][] grid) {
        int nrow = grid.length;
        if (nrow == 0) return 0;
        int ncol = grid[0].length;
        
        int[][] s = new int[nrow][ncol];
        for (int[] row : s)
            Arrays.fill(row, Integer.MAX_VALUE);
        s[0][0] = grid[0][0];

        for (int d = 0; d <= nrow + ncol - 2; ++d) {
            for (int r = 0; r <= d; ++r) {
                int c = d - r;
                if (r >= 0 && r < nrow && c >= 0 && c < ncol) {
                    if (r < nrow - 1 && s[r + 1][c] > s[r][c] + grid[r + 1][c])
                        s[r + 1][c] = s[r][c] + grid[r + 1][c];
                    if (c < ncol - 1 && s[r][c + 1] > s[r][c] + grid[r][c + 1])
                        s[r][c + 1] = s[r][c] + grid[r][c + 1];
                }
            }
        }
        
        return s[nrow - 1][ncol - 1];
    }
}

File: leetcode/Minimum_Path_Sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int> > &grid) {
        int num_row = grid.size();
        if (num_row == 0)
            return 0;
        int num_col = grid[0].size();
        if (num_col == 0)
            return 0;
        vector<vector<int> > min_sum(
            num_row, 
            vector<int>(num_col, numeric_limits<int>::max())
            );
        min_sum[0][0] = grid[0][0];
        for (int i = 0; i < num_row; ++i)
            for (int j = 0; j < num_col; ++j) {
                if (j + 1 < num_col && 
                    min_sum[i][j + 1] > min_sum[i][j] + grid[i][j + 1])
                    min_sum[i][j + 1] = min_sum[i][j] + grid[i][j + 1];
                if (i + 1 < num_row && 
                    min_sum[i + 1][j] > min_sum[i][j] + grid[i + 1][j])
                    min_sum[i + 1][j] = min_sum[i][j] + grid[i + 1][j];
            }
        return min_sum[num_row - 1][num_col - 1];
    }
};
